File: leetcode/Merge-k-Sorted-Lists/Merge-k-Sorted-Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
    /*
     * keep merging the unmerged lists in lists until there is exactly one list remained.
     * do not merge lists[0] with lists[1], lists[2], ..., lists[n - 1] sequentially since that will cause TLE.
	 * merge the first two lists and then push it back to lists and erase the first two to avoid repeated merging.
    */
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()){
            return NULL;
        }
        int n = lists.size();
        while(n > 1){
            for(int i = 0; i < n/2; ++i)
                lists[i] = mergeTwoLists(lists[i], lists[n - 1 - i]);
            n = (n + 1)/2;
        }
        return lists[0];
    }
    
private:
	/* This is the answer:https://leetcode.com/problems/merge-two-sorted-lists/ */
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == NULL) return l2;
        if(l2 == NULL) return l1;
        if(l1 == NULL || l2 == NULL) return NULL;
        ListNode *head = NULL;
        if(l1->val < l2->val){
            head = l1;
            l1 = l1->next;
        }else{
            head = l2;
            l2 = l2->next;
        }
        ListNode *cur = head;
        while(l1 != NULL && l2 != NULL){
            if(l1->val < l2->val){
                cur->next = l1;
                l1 = l1->next;
            }else{
                cur->next = l2;
                l2 = l2->next;
            }
            cur = cur->next;
        }
        if(l1 == NULL){
            cur->next = l2;
        }else{
            cur->next = l1;
        }
        
        return head;
    }
    
};

File: leetcode/python3/merge-k-sorted-lists.py
#
# @lc app=leetcode id=23 lang=python3
#
# [23] Merge k Sorted Lists
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        # # Nlog(N)
        # nums = []
        # for l in lists:
        #     while l:
        #         nums.append(l.val)
        #         l = l.next
        # head = point = ListNode(None)
        # for num in sorted(nums):
        #     point.next = ListNode(num)
        #     point = point.next
        # return head.next


        # Nlog(k)
        if not lists: return
        n = len(lists)
        step = 1
        while step < n:
            for i in range(0, n-step, step*2):
                lists[i] = self.merge2Lists(lists[i], lists[i+step])
            step *= 2
        return lists[0]
        

    def merge2Lists(self, l1, l2):
        head = point = ListNode(None)
        while l1 and l2:
            if l1.val < l2.val:
                point.next = l1
                l1 = l1.next
            else:
                point.next = l2
                l2 = l2.next
            point = point.next
        if l1:
            point.next = l1
        if l2:
            point.next = l2
        return head.next
        

